### PR TITLE
[4.0] Added missing calls to wasmifs on read-only threads

### PR DESCRIPTION
**Pull request by linh2931**
Resolve https://github.com/AntelopeIO/leap/issues/963

- Added missing calls to wasmifs on read-only threads
- Removed first_block_num_used from wasm_instantiation_cache.emplace
- Replaced thread_local for wasmifs with a local variable (unordered_map)

### CODE REVIEW

File: libraries/chain/webassembly/cf_system.cpp [hunk 1]
```diff
@@ -46,6 +46,6 @@ namespace eosio { namespace chain { namespace webassembly {
    }
 
    void interface::eosio_exit( int32_t code ) const {
-      context.control.get_wasm_interface().exit();
+      context.control.wasm_interface_exit();
```
[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/webassembly/cf_system.cpp:49]: Thanks Matt. You are correct. I was wrong to exit other executing contracts. Am reverting it.

[the remark this answers is not in the document]
Comment (linh2931) [reply libraries/chain/webassembly/cf_system.cpp:49]: I am looking into https://github.com/AntelopeIO/leap/issues/956 after this PR.

**Comment by heifner** [comment]
This `static` map causes instability:
https://github.com/AntelopeIO/leap/blob/9a10c803b3a49fd08213ddea11f1d6552a037e9f/libraries/wasm-jit/Source/IR/Types.cpp#L20

Here is how I fixed it:
```
diff --git a/libraries/wasm-jit/Source/IR/Types.cpp b/libraries/wasm-jit/Source/IR/Types.cpp
index d1ade9c3b1..ff3704fd67 100644
--- a/libraries/wasm-jit/Source/IR/Types.cpp
+++ b/libraries/wasm-jit/Source/IR/Types.cpp
@@ -1,6 +1,7 @@
 #include "Types.h"

 #include <map>
+#include <mutex>

 namespace IR
 {
@@ -25,6 +26,8 @@ namespace IR
        template<typename Key,typename Value,typename CreateValueThunk>
        Value findExistingOrCreateNew(std::map<Key,Value>& map,Key&& key,CreateValueThunk createValueThunk)
        {
+      static std::mutex mtx;
+      std::lock_guard<std::mutex> g(mtx);
                auto mapIt = map.find(key);
                if(mapIt != map.end()) { return mapIt->second; }
                else
```
There is tabs in that file, so the indentation looks weird in the diff.

**Review by heifner — CHANGES_REQUESTED** [review]
Fix use of static map

**Comment by spoonincode** [comment]
@heifner that maybe should be another issue; though it wouldn't surprise me if it's related to #969 since what is pointed to in there may be the root cause of WAVM code being touched from multiple threads.

**Review by heifner — APPROVED** [review]
Approved assuming that the `static map` will be fixed in a diff PR.

**Comment by linh2931** [comment]
I am doing a separate PR for https://github.com/AntelopeIO/leap/issues/969 shortly.

**Comment by linh2931** [comment]
The PR is https://github.com/AntelopeIO/leap/pull/975

**Comment by heifner** [comment]
> This `static` map causes instability:

Verified that https://github.com/AntelopeIO/leap/pull/975 fixes this issue.

**Comment by linh2931** [comment]
Thanks @heifner and @spoonincode!

File: libraries/chain/controller.cpp [hunk 2]
```diff
@@ -252,11 +252,9 @@ struct controller_impl {
 #if defined(EOSIO_EOS_VM_RUNTIME_ENABLED) || defined(EOSIO_EOS_VM_JIT_RUNTIME_ENABLED)
    thread_local static vm::wasm_allocator wasm_alloc; // a copy for main thread and each read-only thread
 #endif
-   // Ideally wasmif should be thread_local which must be a static.
-   // Unittests can create multiple controller objects (testers) at the same time,
-   // which overwrites the same static wasmif, is used for eosvmoc too.
    wasm_interface  wasmif;  // used by main thread and all threads for EOSVMOC
-   thread_local static std::unique_ptr<wasm_interface> wasmif_thread_local; // a copy for each read-only thread, used by eos-vm and eos-vm-jit
+   std::mutex threaded_wasmifs_mtx;
+   std::unordered_map<std::thread::id, std::unique_ptr<wasm_interface>> threaded_wasmifs; // one for each read-only thread, used by eos-vm and eos-vm-jit
```
Comment (spoonincode) [inline libraries/chain/controller.cpp:257]: I'm not sure you need a `unique_ptr` here.. couldn't you just `emplace()` the `wasm_interface` in?

**Review by spoonincode — APPROVED** [review]
Looks good

**Pull request merged**